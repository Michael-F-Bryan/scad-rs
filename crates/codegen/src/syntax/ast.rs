use heck::{ToPascalCase, ToShoutySnekCase, ToSnekCase};
use proc_macro2::{Ident, TokenStream};
use quote::{format_ident, quote, ToTokens};

use crate::syntax::{EnumNode, Field, Multiplicity, StructNode, SyntacticElements, Variant};

#[derive(Debug, Clone)]
pub struct Ast {
    structs: Vec<StructNode>,
    enums: Vec<EnumNode>,
}

impl Ast {
    pub(crate) fn new(elements: &SyntacticElements) -> Self {
        let SyntacticElements { enums, structs, .. } = elements;

        Ast {
            enums: enums.clone(),
            structs: structs.clone(),
        }
    }
}

impl ToTokens for Ast {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Ast { enums, structs } = self;

        enums
            .iter()
            .map(|e| e as &dyn ToTokens)
            .chain(structs.iter().map(|s| s as &dyn ToTokens))
            .for_each(|tok| tok.to_tokens(tokens));
    }
}

impl EnumNode {
    fn definition(&self) -> TokenStream {
        let EnumNode { ident, variants } = self;
        let variants = variants.iter().map(|v| {
            let type_name = v.type_name();
            let name = v.name();

            if v.is_token {
                quote!(#name(rowan::api::SyntaxNode<crate::OpenSCAD>))
            } else {
                quote!(#name(#type_name))
            }
        });

        quote! {
            #[derive(Debug, Clone, PartialEq, Eq, Hash)]
            pub enum #ident {
                #(#variants,)*
            }
        }
    }

    fn methods(&self) -> TokenStream {
        let EnumNode { ident, variants } = self;

        let getter = variants.iter().map(|v| {
            let type_name = if v.is_token {
                quote!(rowan::api::SyntaxNode<crate::OpenSCAD>)
            } else {
                v.type_name().to_token_stream()
            };
            let name = v.name();
            let method_name = format_ident!("as_{}", name.to_string().to_snek_case());

            quote! {
                pub fn #method_name(&self) -> Option<#type_name> {
                    match self {
                        #ident :: #name(node) => Some(node.clone()),
                        _ => None,
                    }
                }
            }
        });

        quote! {
            impl #ident {
                #( #getter )*
            }
        }
    }

    fn conversions(&self) -> TokenStream {
        let EnumNode { ident, variants } = self;
        let mut tokens = TokenStream::new();

        for variant in variants {
            if variant.multiplicity != Multiplicity::One || variant.is_token {
                continue;
            }

            let type_name = variant.type_name();
            let name: Ident = variant.name();

            let t = quote! {
                impl From<#type_name> for #ident {
                    fn from(value: #type_name) -> #ident {
                        #ident::#name(value)
                    }
                }
            };
            tokens.extend(t);
        }

        tokens
    }

    fn ast_node_impl(&self) -> TokenStream {
        let EnumNode { ident, variants } = self;

        let can_casts = variants.iter().map(|v| {
            let type_name = v.type_name();

            if v.is_token {
                let kind = v.syntax_kind();
                quote!(kind == #kind)
            } else {
                quote!(<#type_name as rowan::ast::AstNode>::can_cast(kind))
            }
        });

        let try_casts = variants.iter().map(|v| {
            let type_name = v.type_name();

            if !v.is_token {
                quote! {
                    if let Some(node) = <#type_name as rowan::ast::AstNode>::cast(node.clone()) {
                        return Some(node.into());
                    }
                }
            } else {
                let kind = v.syntax_kind();
                quote! {
                    if node.kind() == #kind {
                        return Some(Self::#type_name(node));
                    }
                }
            }
        });

        let names = variants.iter().map(|v| v.name());

        quote! {
            impl rowan::ast::AstNode for #ident {
                type Language = crate::OpenSCAD;

                fn can_cast(kind: crate::SyntaxKind) -> bool {
                    #( #can_casts )||*
                }

                fn cast(node: rowan::api::SyntaxNode<crate::OpenSCAD>) -> Option<Self> {
                    #( #try_casts )*

                    None
                }

                fn syntax(&self) -> &rowan::api::SyntaxNode<crate::OpenSCAD> {
                    match self {
                        #(
                            #ident::#names(node) => node.syntax(),
                        )*
                    }
                }
            }
        }
    }
}

impl ToTokens for EnumNode {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.extend(self.definition());
        tokens.extend(self.methods());
        tokens.extend(self.ast_node_impl());
        tokens.extend(self.conversions());
    }
}

impl StructNode {
    fn definition(&self) -> TokenStream {
        let name = &self.ident;
        quote! {
            #[derive(Debug, Clone, PartialEq, Eq, Hash)]
            pub struct #name(rowan::api::SyntaxNode<crate::OpenSCAD>);
        }
    }

    fn methods(&self) -> TokenStream {
        let ident = &self.ident;

        let getters = self.fields.iter().map(|f| {
            let Field {
                rule_name,
                label,
                multiplicity,
                is_token,
            } = f;

            let mut method_name = label
                .as_deref()
                .map(|s| s.to_snek_case())
                .unwrap_or_else(|| rule_name.to_snek_case());
            if ["use", "let", "for", "if", "else", "true", "false"].contains(&method_name.as_str()) {
                method_name.insert_str(0, "r#");
            }
            let method_name = format_ident!("{method_name}");
            let type_name = format_ident!("{}", rule_name.to_pascal_case());

            if *is_token {
                let token_kind = format_ident!("{}", rule_name.TO_SHOUTY_SNEK_CASE());
                match multiplicity {
                    Multiplicity::Optional => {
                        let method_name = format_ident!("{method_name}_opt");
                        quote! {
                            pub fn #method_name(&self) -> Option<rowan::api::SyntaxToken<crate::OpenSCAD>> {
                                rowan::ast::support::token(&self.0, crate::SyntaxKind::#token_kind)
                            }
                        }
                    }
                    Multiplicity::One => quote! {
                        pub fn #method_name(&self) -> Option<rowan::api::SyntaxToken<crate::OpenSCAD>> {
                                rowan::ast::support::token(&self.0, crate::SyntaxKind::#token_kind)
                        }
                    },
                    Multiplicity::Multiple => {
                        let method_name = format_ident!("{method_name}s");
                        quote! {
                            pub fn #method_name(&self) -> impl Iterator<Item = rowan::api::SyntaxToken<crate::OpenSCAD>> {
                                self.0.children_with_tokens()
                                    .filter_map(|it| it.into_token())
                                    .filter(|tok| tok.kind() == crate::SyntaxKind::#token_kind)
                            }
                        }
                    }
                }
            } else {
                match multiplicity {
                    Multiplicity::Optional => {
                        let method_name = format_ident!("{method_name}_opt");
                        quote! {
                            pub fn #method_name(&self) -> Option<#type_name> {
                                rowan::ast::support::child(&self.0)
                            }
                        }
                    }
                    Multiplicity::One => quote! {
                        pub fn #method_name(&self) -> Option<#type_name> {
                                rowan::ast::support::child(&self.0)
                        }
                    },
                    Multiplicity::Multiple => {
                        let method_name = format_ident!("{method_name}s");
                        quote! {
                            pub fn #method_name(&self) -> impl Iterator<Item = #type_name> {
                                rowan::ast::support::children(&self.0)
                            }
                        }
                    }
                }
            }
        });

        quote! {
            impl #ident {
                #( #getters )*
            }
        }
    }

    fn ast_node_impl(&self) -> TokenStream {
        let StructNode {
            ident, syntax_kind, ..
        } = self;

        quote! {
            impl rowan::ast::AstNode for #ident {
                type Language = crate::OpenSCAD;

                fn can_cast(kind: crate::SyntaxKind) -> bool {
                    kind == crate::SyntaxKind::#syntax_kind
                }

                fn cast(node: rowan::api::SyntaxNode<crate::OpenSCAD>) -> Option<Self> {
                    if #ident::can_cast(node.kind()) {
                        Some(#ident(node))
                    } else {
                        None
                    }
                }

                fn syntax(&self) -> &rowan::api::SyntaxNode<crate::OpenSCAD> {
                    &self.0
                }
            }
        }
    }
}

impl ToTokens for StructNode {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.extend(self.definition());
        tokens.extend(self.methods());
        tokens.extend(self.ast_node_impl());
    }
}

impl Variant {
    fn type_name(&self) -> Ident {
        format_ident!("{}", self.rule_name.to_pascal_case())
    }

    fn syntax_kind(&self) -> TokenStream {
        let variant = format_ident!("{}", self.rule_name.TO_SHOUTY_SNEK_CASE());
        quote!(crate::SyntaxKind::#variant)
    }

    fn name(&self) -> Ident {
        let name = self.label.as_deref().unwrap_or(self.rule_name.as_str());

        format_ident!("{}", name.to_pascal_case())
    }
}
