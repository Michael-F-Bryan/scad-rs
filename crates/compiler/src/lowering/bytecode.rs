use rowan::ast::AstNode;
use scad_bytecode::{Chunk, Instruction, Program};
use scad_syntax::{ast, SyntaxKind, SyntaxNode, SyntaxToken};

use crate::{lowering::Lowering, Diagnostics};

pub(crate) fn compile(db: &dyn Lowering) -> (Program, Diagnostics) {
    let mut diags = Diagnostics::empty();

    let src = db.src();
    let (pkg, d) = db.parse(src);
    diags.extend(d);

    let (chunk, d) = db.compile_package(pkg);
    diags.extend(d);

    let program = Program { chunk };
    (program, diags)
}

pub(crate) fn compile_package(_: &dyn Lowering, ast: ast::Package) -> (Chunk, Diagnostics) {
    let diags = Diagnostics::empty();
    let mut chunk = Chunk::empty();

    for stmt in ast.statements() {
        if stmt
            .syntax()
            .descendants()
            .any(|d| d.kind() == SyntaxKind::ERROR)
        {
            todo!("Handle dodgy statements");
        }
        match stmt {
            ast::Statement::AssignmentStatement(a) => compile_assignment(&mut chunk, a),
            ast::Statement::ModuleInstantiation(m) => compile_module_instantiation(&mut chunk, m),
            ast::Statement::NamedFunctionDefinition(_)
            | ast::Statement::NamedModuleDefinition(_)
            | ast::Statement::Include(_)
            | ast::Statement::Use(_)
            | ast::Statement::IfStatement(_)
            | ast::Statement::ForStatement(_) => todo!(),
        }
    }

    // The top-level statements in a package act as a function, so we need to
    // return to the "caller" (i.e. the runtime on startup or during an import).
    // This return statement will be attached to the last line in the file.
    let line_number = match ast.syntax().children().last() {
        Some(last_child) => line_number(&last_child),
        None => 0,
    };
    chunk.push_instruction(Instruction::Undef, line_number);
    chunk.push_instruction(Instruction::Return, line_number);

    (chunk, diags)
}

fn compile_assignment(chunk: &mut Chunk, a: ast::AssignmentStatement) {
    let a = a.assignment().unwrap();
    let line_number = line_number(a.syntax());

    compile_expr(chunk, a.expr().unwrap());

    let constant = chunk.push_constant(a.ident_token().unwrap().text());
    chunk.push_instruction(Instruction::DefineGlobal(constant), line_number);
}

fn compile_expr(chunk: &mut Chunk, expr: ast::Expr) {
    match expr {
        ast::Expr::Atom(ast::Atom::LiteralExpr(lit)) => compile_literal(chunk, lit),
        ast::Expr::Atom(ast::Atom::FunctionCall(call)) => compile_function_call(chunk, call),
        ast::Expr::Atom(ast::Atom::IndexExpr(_)) => todo!(),
        ast::Expr::Atom(ast::Atom::LookupExpr(lookup)) => compile_lookup(chunk, lookup),
        ast::Expr::ListExpr(list) => compile_list(chunk, list),
        ast::Expr::RangeExpr(_) => todo!(),
        ast::Expr::UnaryExpr(_) => todo!(),
        ast::Expr::TernaryExpr(_) => todo!(),
        ast::Expr::ParenExpr(_) => todo!(),
        ast::Expr::ListComprehensionExpr(_) => todo!(),
        ast::Expr::BinExpr(bin) => compile_bin_expr(chunk, bin),
    }
}

fn compile_list(chunk: &mut Chunk, list: ast::ListExpr) {
    // First we add an empty list to the stack
    chunk.push_instruction(Instruction::CreateList, line_number(list.syntax()));

    for expr in list.exprs() {
        // Then execute each expression
        compile_expr(chunk, expr.clone());
        // and append it to the list
        chunk.push_instruction(Instruction::AddToList, line_number(expr.syntax()));
    }
}

fn compile_bin_expr(chunk: &mut Chunk, bin: ast::BinExpr) {
    let mut exprs = bin.exprs();
    let lhs = exprs.next().expect("There should be at least 1 expression");

    // First we evaluate the LHS and push it onto the stack
    compile_expr(chunk, lhs);

    for (op, rhs) in bin.bin_ops().zip(exprs) {
        // then we push the RHS onto the stack
        compile_expr(chunk, rhs);

        // and execute the binary op
        let line_number = line_number(op.syntax());
        match op {
            ast::BinOp::Plus(_) => chunk.push_instruction(Instruction::Add, line_number),
            ast::BinOp::Minus(_) => chunk.push_instruction(Instruction::Sub, line_number),
            ast::BinOp::Star(_) => chunk.push_instruction(Instruction::Mul, line_number),
            ast::BinOp::Slash(_) => chunk.push_instruction(Instruction::Div, line_number),
            ast::BinOp::Percent(_) => todo!(),
            ast::BinOp::Caret(_) => todo!(),
            ast::BinOp::GreaterThanEquals(_) => todo!(),
            ast::BinOp::GreaterThan(_) => todo!(),
            ast::BinOp::DoubleEquals(_) => todo!(),
            ast::BinOp::NotEqual(_) => todo!(),
            ast::BinOp::LessThanEquals(_) => todo!(),
            ast::BinOp::LessThan(_) => todo!(),
            ast::BinOp::And(_) => todo!(),
            ast::BinOp::Or(_) => todo!(),
        }
    }
}

fn compile_lookup(chunk: &mut Chunk, lookup: ast::LookupExpr) {
    let line_number = line_number(lookup.syntax());
    let path: Vec<_> = lookup.ident_tokens().collect();

    let name = match path.as_slice() {
        [name] => name.text(),
        _ => todo!("Handle dotted paths ({})", lookup.syntax().to_string()),
    };

    let constant = chunk.push_constant(name);
    chunk.push_instruction(Instruction::LookupVariable(constant), line_number);
}

fn compile_literal(chunk: &mut Chunk, lit: ast::LiteralExpr) {
    let line_number = line_number(lit.syntax());

    match lit {
        ast::LiteralExpr::TrueKw(_) => chunk.push_instruction(Instruction::True, line_number),
        ast::LiteralExpr::FalseKw(_) => chunk.push_instruction(Instruction::False, line_number),
        ast::LiteralExpr::UndefKw(_) => chunk.push_instruction(Instruction::Undef, line_number),
        ast::LiteralExpr::Integer(node) | ast::LiteralExpr::Float(node) => {
            let number: f64 = node.first_token().unwrap().text().parse().unwrap();
            let constant = chunk.push_constant(number);
            chunk.push_instruction(Instruction::Constant(constant), line_number);
        }
        ast::LiteralExpr::String(node) => {
            let constant = chunk.push_constant(node.first_token().unwrap().text());
            chunk.push_instruction(Instruction::Constant(constant), line_number);
        }
    }
}

fn line_number(_node: &SyntaxNode) -> u16 {
    // TODO: Actually calculate the line number (probably via Salsa so we get
    // caching)
    0
}

fn compile_call(
    chunk: &mut Chunk,
    ident: SyntaxToken,
    args: Option<ast::Arguments>,
    node: &SyntaxNode,
) {
    // When you execute a function/module, first you evaluate its arguments...
    let args: Vec<_> = args.into_iter().flat_map(|a| a.arguments()).collect();
    let arg_count = args.len();

    for arg in args {
        match arg {
            ast::Argument::Expr(expr) => compile_expr(chunk, expr),
            ast::Argument::Assignment(_) => todo!(),
        }
    }

    // ... then you push the function object onto the stack
    let line_number = line_number(node);
    let constant = chunk.push_constant(ident.text());
    chunk.push_instruction(Instruction::LookupVariable(constant), line_number);

    // and finally, we can emit the call
    chunk.push_instruction(
        Instruction::Call(
            arg_count
                .try_into()
                .expect("Function calls don't support more than 255 arguments"),
        ),
        line_number,
    );
}

fn compile_module_instantiation(chunk: &mut Chunk, m: ast::ModuleInstantiation) {
    compile_call(
        chunk,
        m.ident_token().unwrap(),
        m.arguments_opt(),
        m.syntax(),
    );
    chunk.push_instruction(Instruction::SaveGeometry, line_number(m.syntax()));
}

fn compile_function_call(chunk: &mut Chunk, f: ast::FunctionCall) {
    compile_call(
        chunk,
        f.ident_token().unwrap(),
        f.arguments_opt(),
        f.syntax(),
    )
}

#[cfg(test)]
mod tests {
    use scad_bytecode::Disassembler;

    use crate::{db::Database, parsing::Parsing};

    use super::*;

    macro_rules! bytecode_test {
        ($name:ident, $src:literal) => {
            #[test]
            fn $name() {
                let mut db = Database::default();
                let src = $src;
                db.set_src(src.into());
                eprintln!("---- Source ----");
                eprintln!("{src}");

                eprintln!("---- AST ----");
                let (ast, _) = db.parse(src.into());
                eprintln!("{ast:#?}");

                eprintln!("---- Program ----");
                let (program, diags) = db.compile();
                eprintln!("{program:#?}");

                eprintln!("---- Disassembly ----");
                let mut dis = Disassembler::default();
                dis.program(&program);
                let disassembly = dis.finish();
                eprintln!("{disassembly}");

                eprintln!("---- Errors ----");
                eprintln!("{diags:#?}");

                assert!(!diags.has_warnings());

                insta::with_settings! {
                    {
                        description => $src,
                        omit_expression => true,
                    },
                    {
                        insta::assert_display_snapshot!(disassembly);
                    }
                }
            }
        };
    }

    bytecode_test!(
        print_addition_with_variables,
        "x = 42; y = x + 2; print(x);"
    );
    bytecode_test!(
        norm_tests,
        "u=undef; echo(norm([])); echo(norm([1, 2, [1, 3]]));"
    );
}
