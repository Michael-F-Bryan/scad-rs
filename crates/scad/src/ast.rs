//! The *Abstract Syntax Tree* representation of a program.
//!
//! Every program lives in a [`Package`].

use std::str::FromStr;

use rowan::{ast::AstNode, SyntaxNode, SyntaxText, TextSize};

use crate::syntax::{OpenSCAD, SyntaxKind};

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Package(SyntaxNode<OpenSCAD>);

impl Package {
    pub fn statements(&self) -> impl Iterator<Item = Statement> {
        self.0.children().filter_map(Statement::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Statement(SyntaxNode<OpenSCAD>);

impl Statement {
    pub fn as_assignment(&self) -> Option<Assignment> {
        self.0.children().find_map(Assignment::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Assignment(SyntaxNode<OpenSCAD>);

impl Assignment {
    pub fn identifier(&self) -> Option<Ident> {
        self.0.children().find_map(|c| Ident::cast(c))
    }

    pub fn expression(&self) -> Option<Expr> {
        self.0.children().find_map(|c| Expr::cast(c))
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Ident(SyntaxNode<OpenSCAD>);

impl Ident {
    pub fn name(&self) -> SyntaxText {
        self.0.text()
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Expr(SyntaxNode<OpenSCAD>);

impl Expr {
    pub fn number(&self) -> Option<Number> {
        self.0.children().find_map(Number::cast)
    }

    pub fn boolean(&self) -> Option<Boolean> {
        self.0.children().find_map(Boolean::cast)
    }

    pub fn string(&self) -> Option<StringLiteral> {
        self.0.children().find_map(StringLiteral::cast)
    }

    pub fn undef(&self) -> Option<Undefined> {
        self.0.children().find_map(Undefined::cast)
    }

    pub fn binary_op(&self) -> Option<(Expr, SyntaxNode<OpenSCAD>, Expr)> {
        let useful_nodes: Vec<_> = self
            .0
            .children()
            .filter(|c| match c.kind() {
                SyntaxKind::COMMENT | SyntaxKind::WHITESPACE => false,
                _ => true,
            })
            .collect();

        let [left, op, right]: [SyntaxNode<OpenSCAD>; 3] = dbg!(useful_nodes.try_into().ok()?);

        let left = Expr::cast(left)?;
        let right = Expr::cast(right)?;

        Some((left, op, right))
    }

    pub fn lookup(&self) -> Option<Lookup> {
        self.0.children().find_map(Lookup::cast)
    }

    pub fn function_call(&self) -> Option<FunctionCall> {
        self.0.children().find_map(FunctionCall::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Lookup(SyntaxNode<OpenSCAD>);

impl Lookup {
    pub fn ident(&self) -> Option<Ident> {
        self.0.children().find_map(Ident::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct FunctionCall(SyntaxNode<OpenSCAD>);

impl FunctionCall {
    pub fn ident(&self) -> Option<Ident> {
        self.0.children().find_map(Ident::cast)
    }

    pub fn parameters(&self) -> Option<Parameters> {
        self.0.children().find_map(Parameters::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Parameters(SyntaxNode<OpenSCAD>);

impl Parameters {
    pub fn iter(&self) -> impl Iterator<Item = Parameter> {
        self.0.children().filter_map(Parameter::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Parameter(SyntaxNode<OpenSCAD>);

impl Parameter {
    pub fn named(&self) -> Option<Assignment> {
        self.0.children().find_map(Assignment::cast)
    }

    pub fn positional(&self) -> Option<Expr> {
        self.0.children().find_map(Expr::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Number(SyntaxNode<OpenSCAD>);

impl Number {
    pub fn literal(&self) -> SyntaxText {
        self.0.text()
    }

    pub fn parse<T: FromStr>(&self) -> Result<T, T::Err> {
        let repr = self.literal().to_string();
        repr.parse()
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Boolean(SyntaxNode<OpenSCAD>);

impl Boolean {
    pub fn literal(&self) -> SyntaxText {
        self.0.text()
    }

    pub fn value(&self) -> bool {
        match self.0.kind() {
            SyntaxKind::TRUE_KW => true,
            SyntaxKind::FALSE_KW => false,
            _ => unreachable!(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct StringLiteral(SyntaxNode<OpenSCAD>);

impl StringLiteral {
    pub fn literal(&self) -> SyntaxText {
        self.0.text()
    }

    pub fn value(&self) -> String {
        let full_range = self.0.text_range();

        self.literal()
            .slice(TextSize::from(1)..full_range.len() - TextSize::from(1))
            .to_string()
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Undefined(SyntaxNode<OpenSCAD>);

macro_rules! impl_ast_node {
    ($(
        $name:ident => $kind:pat
    ),* $(,)?) => {
        $(

            impl AstNode for $name {
                type Language = OpenSCAD;

                fn can_cast(kind: <Self::Language as rowan::Language>::Kind) -> bool
                where
                    Self: Sized,
                {
                    match kind {
                        $kind => true,
                        _ => false,
                    }
                }

                fn cast(node: SyntaxNode<Self::Language>) -> Option<Self>
                where
                    Self: Sized,
                {
                    if $name::can_cast(node.kind()) {
                        Some($name(node))
                    } else {
                        None
                    }
                }

                fn syntax(&self) -> &SyntaxNode<Self::Language> {
                    &self.0
                }
            }
        )*
    };
}

impl_ast_node! {
    Package => SyntaxKind::PACKAGE,
    Ident => SyntaxKind::IDENTIFIER,
    Expr => SyntaxKind::EXPR,
    Assignment => SyntaxKind::ASSIGNMENT,
    Number => SyntaxKind::NUMBER_LIT,
    Boolean => SyntaxKind::TRUE_KW | SyntaxKind::FALSE_KW,
    StringLiteral => SyntaxKind::STRING_LIT,
    Undefined => SyntaxKind::UNDEF_KW,
    Statement => SyntaxKind::STATEMENT,
    Lookup => SyntaxKind::LOOKUP,
    FunctionCall => SyntaxKind::FUNCTION_CALL,
    Parameters => SyntaxKind::PARAMETERS,
    Parameter => SyntaxKind::PARAMETER,
}
