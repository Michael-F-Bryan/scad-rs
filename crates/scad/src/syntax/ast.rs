use std::str::FromStr;

use rowan::{ast::AstNode, SyntaxNode, SyntaxText, TextRange, TextSize};

use crate::syntax::{lexer::OpenSCAD, SyntaxKind};

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Package(SyntaxNode<OpenSCAD>);

impl Package {
    pub fn statements(&self) -> impl Iterator<Item = Statement> {
        self.0.children().filter_map(Statement::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Statement(SyntaxNode<OpenSCAD>);

impl Statement {
    pub fn as_assignment(&self) -> Option<Assignment> {
        self.0.children().find_map(Assignment::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Assignment(SyntaxNode<OpenSCAD>);

impl Assignment {
    pub fn identifier(&self) -> Option<Ident> {
        self.0.children().find_map(|c| Ident::cast(c))
    }

    pub fn expression(&self) -> Option<Expr> {
        self.0.children().find_map(|c| Expr::cast(c))
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Ident(SyntaxNode<OpenSCAD>);

impl Ident {
    pub fn name(&self) -> SyntaxText {
        self.0.text()
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Expr(SyntaxNode<OpenSCAD>);

impl Expr {
    pub fn number(&self) -> Option<Number> {
        self.0.children().find_map(Number::cast)
    }

    pub fn boolean(&self) -> Option<Boolean> {
        self.0.children().find_map(Boolean::cast)
    }

    pub fn string(&self) -> Option<StringLiteral> {
        self.0.children().find_map(StringLiteral::cast)
    }

    pub fn undef(&self) -> Option<Undefined> {
        self.0.children().find_map(Undefined::cast)
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Number(SyntaxNode<OpenSCAD>);

impl Number {
    pub fn literal(&self) -> SyntaxText {
        self.0.text()
    }

    pub fn parse<T: FromStr>(&self) -> Result<T, T::Err> {
        let repr = self.literal().to_string();
        repr.parse()
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Boolean(SyntaxNode<OpenSCAD>);

impl Boolean {
    pub fn literal(&self) -> SyntaxText {
        self.0.text()
    }

    pub fn value(&self) -> bool {
        match self.0.kind() {
            SyntaxKind::TRUE_KW => true,
            SyntaxKind::FALSE_KW => false,
            _ => unreachable!(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct StringLiteral(SyntaxNode<OpenSCAD>);

impl StringLiteral {
    pub fn literal(&self) -> SyntaxText {
        self.0.text()
    }

    pub fn value(&self) -> String {
        let full_range = self.0.text_range();

        let range = TextRange::new(
            full_range.start() + TextSize::from(1),
            full_range.end() - TextSize::from(1),
        );
        self.literal().slice(range).to_string()
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub struct Undefined(SyntaxNode<OpenSCAD>);

macro_rules! impl_ast_node {
    ($(
        $name:ident => $kind:pat
    ),* $(,)?) => {
        $(

            impl AstNode for $name {
                type Language = OpenSCAD;

                fn can_cast(kind: <Self::Language as rowan::Language>::Kind) -> bool
                where
                    Self: Sized,
                {
                    match kind {
                        $kind => true,
                        _ => false,
                    }
                }

                fn cast(node: SyntaxNode<Self::Language>) -> Option<Self>
                where
                    Self: Sized,
                {
                    if $name::can_cast(node.kind()) {
                        Some($name(node))
                    } else {
                        None
                    }
                }

                fn syntax(&self) -> &SyntaxNode<Self::Language> {
                    &self.0
                }
            }
        )*
    };
}

impl_ast_node! {
    Package => SyntaxKind::PACKAGE,
    Ident => SyntaxKind::IDENTIFIER,
    Expr => SyntaxKind::EXPR,
    Assignment => SyntaxKind::ASSIGNMENT,
    Number => SyntaxKind::NUMBER_LIT,
    Boolean => SyntaxKind::TRUE_KW | SyntaxKind::FALSE_KW,
    StringLiteral => SyntaxKind::STRING_LIT,
    Undefined => SyntaxKind::UNDEF_KW,
    Statement => SyntaxKind::STATEMENT,
}
