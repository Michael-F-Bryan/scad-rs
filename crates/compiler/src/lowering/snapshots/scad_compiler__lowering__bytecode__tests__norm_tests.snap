---
source: crates/compiler/src/lowering/bytecode.rs
description: "u=undef; echo(norm([])); echo(norm([1, 2, [1, 3]]));"
---
  0|  undef
   |  define_global u
   |  create_list
   |  lookup norm
   |  call (1 arg)
   |  lookup echo
   |  call (1 arg)
   |  save_geometry
   |  create_list
   |  load_constant 1
   |  add_to_list
   |  load_constant 2
   |  add_to_list
   |  create_list
   |  load_constant 1
   |  add_to_list
   |  load_constant 3
   |  add_to_list
   |  add_to_list
   |  lookup norm
   |  call (1 arg)
   |  lookup echo
   |  call (1 arg)
   |  save_geometry
   |  ret

